src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Services/PackageSearchMetadataCacheItem.cs

```diff
@@ -0,0 +1,76 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System.Collections.Generic;
+using System.Globalization;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft;
+using NuGet.Packaging.Core;
+using NuGet.Protocol.Core.Types;
+using NuGet.Shared;
+using NuGet.Versioning;
+using NuGet.VisualStudio.Internal.Contracts;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    internal sealed class PackageSearchMetadataCacheItem
+    {
+        private IPackageSearchMetadata _packageSearchMetadata;
+        private readonly IPackageMetadataProvider _packageMetadataProvider;
+        private readonly Dictionary<NuGetVersion, PackageSearchMetadataCacheItemEntry> _cachedItemEntries;
+
+        public PackageSearchMetadataCacheItem(IPackageSearchMetadata packageSearchMetadata, IPackageMetadataProvider packageMetadataProvider)
+        {
+            _cachedItemEntries = new Dictionary<NuGetVersion, PackageSearchMetadataCacheItemEntry>
+            {
+                [packageSearchMetadata.Identity.Version] = new PackageSearchMetadataCacheItemEntry(packageSearchMetadata, packageMetadataProvider)
+            };
+
+            _packageSearchMetadata = packageSearchMetadata;
+            _packageMetadataProvider = packageMetadataProvider;
+            AllVersionsContextInfo = GetVersionInfoContextInfoAsync();
+        }
+
+        public ValueTask<IReadOnlyCollection<VersionInfoContextInfo>> AllVersionsContextInfo { get; private set; }
+
+        public async ValueTask<PackageSearchMetadataCacheItemEntry> GetPackageSearchMetadataCacheVersionedItemAsync(PackageIdentity packageIdentity, CancellationToken cancellationToken)
+        {
+            Assumes.NotNull(packageIdentity);
+
+            if (!_cachedItemEntries.TryGetValue(packageIdentity.Version, out PackageSearchMetadataCacheItemEntry cacheItemEntry))
+            {
+                IPackageSearchMetadata packageSearchMetadata = await _packageMetadataProvider.GetPackageMetadataAsync(packageIdentity, true, cancellationToken);
+                cacheItemEntry = new PackageSearchMetadataCacheItemEntry(packageSearchMetadata, _packageMetadataProvider);
+                _cachedItemEntries[packageIdentity.Version] = cacheItemEntry;
```

Comment:
Do we need thread safety here?

Reply:
Yes we do, will switch to ConcurrentDictionary